qiskit/algorithms/optimizers/steppable_optimizer.py

```diff
@@ -0,0 +1,213 @@
+    njev: Optional[int]
+    nit: Optional[int]
+
+
+class SteppableOptimizer(Optimizer):
+    """
+    Base class for a steppable optimizer.
+
+    This family of optimizers will be using the ask and tell interface.
+    When using this interface the user has to call the function ask() in order to get information about how to evaluate the fucntion
+    (we are asking the optimizer about how to do the evaluation). This information will be mostly about at what point should we evaluate
+    the function next, but depending on the optimizer it can also be about whether we should evaluate the function itself or its gradient.
+
+    Once the function has been evaluated, the user calls the method tell() to tell the optimizer what has been the result of the function evaluation.
+    The optimizer then updates its state accordingly and the user can decide whether to stop the optimization process or to repeat a step.
+
+    This interface is more customizable, and allows the user to have full control over the evaluation of the function.
+
+    For example:
+    .. code-block::python
+
+        def objective(x):
+            if random.choice([True, False]):
+                return None
+            else:
+                return (np.linalg.norm(x) - 1) ** 2
+
+        def grad(x):
+            if random.choice([True, False]):
+                return None
+            else:
+                return 2 * (np.linalg.norm(x) - 1) * x / np.linalg.norm(x)
+
+
+        initial_point = np.random.normal(0, 1, size=(100,))
+
+        optimizer = SteppableGradientDescent(maxiter=20)
+        optimizer.initialize(x0=initial_point, fun=objective, jac=grad)
+
+        for _ in range(20):
+            ask_object = optimizer.ask()
+            evaluated_gradient = None
+
+            while evaluated_gradient is None:
+                evaluated_gradient = grad(ask_object.x_center)
+                optimizer._state.njev += 1
+
+            tell_object = GD_TellObject(gradient=evaluated_gradient)
+            optimizer.tell(ask_object=ask_object, tell_object=tell_object)
+
+        result = optimizer.create_result()
+
+    In this case the evaluation of the function has a chance of failing. The user, with specific knowledge about his function can catch this errors and handle before
+    passing the result to the optimizer.
+
+    In case the user isn't dealing with complicated function and is more familiar with step by step optimization algorithms, a method step() has been created that
+    acts as a wrapper for ask() and tell().
+    In the same spirit the method minimize() will optimize the function and return the result without the user having to worry about the optimization process.
+
+    To see other libraries that use this interface one can visit: https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/009_ask_and_tell.html
+
+
+    """
+
+    def __init__(
+        self,
+        callback: Optional[CALLBACK] = None,
+        maxiter: int = None,
+
+    ):
```

Comment:
We usually put the `maxiter` as first argument since it's more often specified than the callback 🙂 
```suggestion
        maxiter: Optional[int] = None,
        callback: Optional[CALLBACK] = None,
    ) -> None:
```